R/utilities-simulation-results.R

```diff
@@ -0,0 +1,132 @@
+#'
+#' @param simulationResults Object of type \code{SimulationResults} produced by calling \code{runSimulation}
+#' on a \code{Simulation} object.
+#' @param quantitiesOrPaths Quantity instances (element or vector) (typically retrieved using
+#' \code{getAllQuantitiesMatching}) or quantity path (element or vector of strings) for which the results are to be returned.
+#' When providing the paths, only absolute full paths are supported (i.e., no matching with '*' possible)
+#' @param individualIds \code{numeric} IDs of individiuals for which the results should be extracted.
+#' By default, all individuals from the results are considered. If the individual with the provided ID is not found, the ID is ignored
+#'
+#' @return A list of results, each entry having the name of a quantity path and consisting of the element \code{x} being the time values (in minutes)
+#' and the element \code{y} a \code{NxM} array of simulated values, with \code{N} the number of entries (equals the number of output time steps)
+#' and \code{M} the number of individuals. If no individual contains the results for a given path, the value of the element is \code{NULL}
+#'
+#' @seealso \code{\link{getResultsValues}}
+#'
+#' @export
+#'
+#' @examples
+#' simPath <- system.file("extdata", "simple.pkml", package = "ospsuite")
+#' sim <- loadSimulation(simPath)
+#' simResults <- runSimulation(sim)
+#'
+#' paths <- c(
+#'   "Organism|PeripheralVenousBlood|Caffeine|Plasma (Peripheral Venous Blood)",
+#'   "Organism|ArterialBlood|Plasma|Caffeine|Concentration in container"
+#' )
+#'
+#' resultsValues <- getPopulationResultsValues(simResults, paths, individualIds = 0)
+#' # plot(resultsValues$`Organism|PeripheralVenousBlood|Caffeine|Plasma (Peripheral Venous Blood)`)
+getPopulationResultsValues <- function(simulationResults, quantitiesOrPaths, individualIds = NULL) {
+  validateIsOfType(simulationResults, "SimulationResults")
+  quantitiesOrPaths <- c(quantitiesOrPaths)
+  validateIsOfType(quantitiesOrPaths, c("Quantity", "character"))
+  validateIsNumeric(individualIds, nullAllowed = TRUE)
+
+  # If quantities are passed, get their paths.
+  paths <- quantitiesOrPaths
+  if (isOfType(quantitiesOrPaths, "Quantity")) {
+    paths <- lapply(quantitiesOrPaths, function(x) x$path)
+  }
+  paths <- unique(paths)
+
+  # If no specific individual ids are passed, iterate through all individuals
+  if (is.null(individualIds)) {
+    nrOfIndividuals <- simulationResults$count
+    # TODO: Can we assume that all individuals ids are ascending number from 0 to the number of individuals-1??
+    individualIds <- 0:(nrOfIndividuals - 1)
+  }
+  else {
+    individualIds <- unique(individualIds)
+    nrOfIndividuals <- length(individualIds)
+  }
+
+  # Pre-alloate a list with the number of entries equal to the number of paths
+  output <- vector("list", length(paths))
+  names(output) <- paths
+  # All time values are equal
+  timeVals <- simulationResults$timeValues
+
+  for (path in paths) {
```

Comment:
The output contains entries for all paths (or quantities) that were provided - even if they are not in the results! I thing this is fair since the user might try to access them, and it is better to return a `NULL` rather then just to completely remove the entry from the outputs...